src/Data/Singletons/Promote.hs

```diff
@@ -313,10 +305,21 @@ promoteLetDecs mb_let_uniq decls = do
 --    type level in the same way they can be used uniformly at the type level.
 --
 --  * for each nullary data constructor we generate a type synonym
-promoteDataDec :: DataDecl -> PrM ()
-promoteDataDec (DataDecl _name _tvbs ctors) = do
-  ctorSyms <- buildDefunSymsDataD ctors
-  emitDecs ctorSyms
+--
+--  * we create a "manual" record selector function and return its DLetDecs so
+--    that they can be promoted later.
+--    See Note [singletons and record selectors] in D.S.Single.Data.
```

Comment:
So this emits declarations for the promoted fixity declarations for record selectors, but just returns the record selector simulated functions unpromoted? If that simplifies the plumbing, then sure, but it should be carefully commented.

Reply:
You raise a good point: the name "`promoteDataDec`" doesn't actually promote a data declaration (not even its record selectors), which makes its name rather misleading. We should name it something else.

Moreover, I'm not sure that `promoteDataDec` is the best place to put that call to `promoteReifiedInfixDecls`, since it would make more sense to put it in `promoteLetDecRHS` (which already does 90% of record selector promotion). Perhaps we should do the remaining 10% there as well.

I'll experiment with these ideas and push a new design when it's ready.

Reply:
Hm, the name "`promoteDataDec`" isn't completely misleading after all, since it also defunctionalizes the data constructors.

Moreover, my plan of moving the invocation of `promoteReifiedInfixDecls` into `promoteLetDecRHS` won't work out cleanly, since `promoteLetDecRHS` is also on each class method whenever an instance is promoted. When promoting an instance method, the `Name` passed to `promoteLetDecRHS` happens to be the name of the method itself. If that name happens to have a fixity declaration in the module that defined it, then `promoteReifiedInfixDecls` will generate an _additional_ promoted fixity declaration every single time that method is defined in an instance! Ugh. We could imagine adding a special case in `promoteLetDecRHS` whenever we are dealing with an instance, but that solution doesn't seem much better than just keeping `promoteReifiedInfixDecls` in `promoteDataDec`.

Bottom line: I have decided to keep the name "`promoteDataDec`" and keep its implementation just the way it is. I have rewritten the documentation for `promoteDataDec`, both to clarify why it returns an unpromoted `DLetDec` and because some of its documentation was irrelevant (e.g., it mentions deriving `Eq` instances, even though `promoteDataDec` has nothing to do with that).